lib/ansible/modules/cloud/docker/docker_container.py

```diff
@@ -1238,6 +1265,76 @@ def running(self):
                 return True
         return False
 
+    def _compare_dict_allow_more_present(self, av, bv):
+        '''
+        Compare two dictionaries for whether every entry of the first is in the second.
+        '''
+        for key, value in av.items():
+            if key not in bv:
+                return False
+            if bv[key] != value:
+                return False
+        return True
+
+    def _compare(self, a, b, compare):
+        '''
+        Compare values a and b as described in compare.
+        '''
+        method = compare['comparison']
+        if method == 'igore':
+            return True
+        # If a or b is None:
+        if a is None or b is None:
+            # If both are None: equality
+            if a == b:
+                return True
+            # Otherwise, not equal for values, and equal
+            # if the other is empty for set/list/dict
+            if compare['type'] == 'value':
+                return False
+            return len(b if a is None else a) == 0
+        # Do proper comparison (both objects not None)
+        if compare['type'] == 'value':
+            return a == b
+        elif compare['type'] == 'list':
+            if method == 'strict':
+                return a == b
+            else:
+                set_a = set(a)
+                set_b = set(b)
+                return set_b >= set_a
+        elif compare['type'] == 'dict':
+            if method == 'strict':
+                return a == b
+            else:
+                return self._compare_dict_allow_more_present(a, b)
+        elif compare['type'] == 'set':
+            set_a = set(a)
+            set_b = set(b)
+            if method == 'strict':
+                return set_a == set_b
+            else:
+                return set_b >= set_a
+        elif compare['type'] == 'set(dict)':
```

Comment:
Not really, see above.

Reply:
Ah, maybe `list(dict)` will be a more suitable type name for this.

And we can add a shortcut for `len(a) == len(b)`, which I think will make the logic a bit clear, like this:
```python
if len(a) != len(b) and method == 'strict':
   return False
for av in a:
    for bv in b:
       # the compare logic, no need to another iteration for bv in b and av in a

Reply:
Well, but it is not a list of dicts, but a set of dicts, since the order of the dicts does not matter (and actually *must* be ignored).

The shortcut won't work because the same entry could show up multiple times in the same list. (This probably will result in an error when creating a new container, but for checking an existing one this would be ok.) I'll add a comment explaining this.

Reply:
And this is not about the order, as we're comparing it in O(n^2), ordering does not matter here.

Reply:
Yep; the original version had `set_a = set(a)` etc., which I changed once I added tests for these options and noticed that it breaks ;)

We could also try to get rid of O(n^2), but that's more complicated and probably even slower, since there will almost never be cases where n is large enough to make a difference.

Reply:
So maybe `list(dict)` is a better match for the current implementation, and then we can use `len(a) == len(b)` shortcut safely.

Reply:
But it is not a list, it is a set (conceptually), since the order doesn't matter.